### PR TITLE
perf: optimize `parseHeaders`

### LINKED ISSUE #2491: Request to improve the performance of parseHeaders
## Bug Description

parseHeaders in the utils.js seems too slow.
## Reproducible By

```js
const dispatcher = new Pool('http://127.0.0.1:8090' /* an 4worker nginx */, {
  pipelining: 30,
  connections: 100,
});

const proxy = http.createServer((req, res) => {
  dispatcher.request({ // the constructor of RequestHandler in undici/lib/api/api-request.js
    method: req.method,
    path: req.url,
    headers: req.headers,
    body: req,
  }, (err, { statusCode, headers, body /** a Readable */ }) => {
    if (err) {
      throw err
    }
    res.writeHead(statusCode, headers);
     pipe(body, res);
  });
});

function pipe(src, dst) {
  src.on('readable', () => {
    let chunk;
    while (null !== (chunk = src.read(65536))) {
      dst.write(chunk);
    }
  });
  src.on('end', () => {
    dst.end()
  });
}
```

## Expected Behavior

Less CPU time spent in parseHeaders

## Logs & Screenshots

image

## Environment

Node 22
linux

### Additional context

I open this issue because I've spotted some optimization for lowercasing the headers in node's built-in HTTP module, https://github.com/nodejs/node/blob/2e458d973638d01fcb6a0d7d611e0120a94f4d35/lib/_http_incoming.js#L279C3-L279C3
basically, it matches the known header field and use the cached string if found. Maybe we can adapt the same trick here to eliminate the `toString()` and `toLowercase()` overhead in parseHeaders.

### PR DESCRIPTION
**Pull request by tsctx**
Fixes #2491 

## benchmark
```js
const { parseHeaders } = require("./undici/lib/core/util")
const benchmark = require("benchmark");

const suite = new benchmark.Suite();

const headers = Object.entries({
  "Content-Type": "application/json",
  Date: "Wed, 01 Nov 2023 00:00:00 GMT",
  "Powered-By": "NodeJS",
  "Content-Encoding": "gzip",
  "Set-Cookie": "__Secure-ID=123; Secure; Domain=example.com",
  "Content-Length": "150",
  Vary: "Accept-Encoding, Accept, X-Requested-With",
}).flat()

suite.add("parseHeaders", () => {
  parseHeaders(headers)
})
suite.on("cycle", function (event) {
  console.log(String(event.target));
});

// run async
new Promise((resolve) => setTimeout(resolve, 7000)).then(() =>
  suite.run({ async: true })
);
```

-  *main*
```
parseHeaders x 754,545 ops/sec ±4.92% (73 runs sampled)
```

- *this patch*
```
parseHeaders x 1,718,324 ops/sec ±3.51% (85 runs sampled)
```

### CODE REVIEW

**Review by ronag — COMMENTED** [review]
Benchmark?

**Comment by tsctx** [comment]
@ronag added

**Review by ronag — APPROVED** [review]
Good job

**Comment by tsctx** [comment]
> Is a Map faster?
Not much change. But, in my past experience, the object was fast if the key was present, but slow if not.
What do you think?
```
parseHeaders x 1,707,785 ops/sec ±2.17% (84 runs sampled)
```

**Comment by tsctx** [comment]
@ronag 
> Is a Map faster?

Not much change.
```
parseHeaders x 1,707,785 ops/sec ±2.17% (84 runs sampled)
```

File: lib/core/util.js [hunk 1]
```diff
@@ -223,19 +224,20 @@ function parseHeaders (headers, obj = {}) {
...
     if (!val) {
       if (Array.isArray(headers[i + 1])) {
-        obj[key] = headers[i + 1].map(x => x.toString('utf8'))
+        obj[lowerCasedKey] = headers[i + 1].map(x => x.toString('utf8'))
       } else {
-        obj[key] = headers[i + 1].toString('utf8')
+        obj[lowerCasedKey] = headers[i + 1].toString('utf8')
       }
     } else {
       if (!Array.isArray(val)) {
```
Comment (tsctx) [inline lib/core/util.js:238]: I think `typeof val === "string"` can be used for this part. Like this.
```js
function parseHeaders (headers, obj) {
  // For H2 support
  if (!Array.isArray(headers)) return headers

  if (obj === undefined) {
    for (let i = 0; i < headers.length; i += 2) {
      const key = headers[i].toString()
      const lowerCasedKey = headerNameLowerCasedRecord[key] ?? key.toLowerCase()
      let val = obj[lowerCasedKey]
  
      if (!val) {
        const headersValue = headers[i + 1]
        if (Array.isArray(headersValue)) {
          obj[lowerCasedKey] = headersValue.map(x => x.toString('utf8'))
        } else {
          obj[lowerCasedKey] = headersValue.toString('utf8')
        }
      } else {
        if (typeof val === "string") {
          val = [val]
          obj[lowerCasedKey] = val
        }
        val.push(headers[i + 1].toString('utf8'))
      }
    }
  } else {
    for (let i = 0; i < headers.length; i += 2) {
      const key = headers[i].toString()
      const lowerCasedKey = headerNameLowerCasedRecord[key] ?? key.toLowerCase()
      let val = obj[lowerCasedKey]
  
      if (!val) {
        const headersValue = headers[i + 1]
        if (Array.isArray(headersValue)) {
          obj[lowerCasedKey] = headersValue.map(x => x.toString('utf8'))
        } else {
          obj[lowerCasedKey] = headersValue.toString('utf8')
        }
      } else {
        if (!Array.isArray(val)) {
          val = [val]
          obj[lowerCasedKey] = val
        }
        val.push(headers[i + 1].toString('utf8'))
      }
    }
  }
  // See https://github.com/nodejs/node/pull/46528
  if ('content-length' in obj && 'content-disposition' in obj) {
    obj['content-disposition'] = Buffer.from(obj['content-disposition']).toString('latin1')
  }

  return obj
}

```

**Review by mcollina — APPROVED** [review]
lgtm

Very good job!

**Pull request merged**